add --enable-sse option

glensc opened the pull request:
add --enable-sse option. #263

adds the starting ground, by default the option is not enabled.

my x86-64 compiler was able to compile with the flags although i'm sure those are not needed on x86-64 target.

ktsaou commented:
will autoconf-archive be needed by users too?

glensc commented:
of course not. you will be pubishing release tarballs that don't need *ANY* auto* package at all

ktsaou commented:
I mean the people installing it from git. netdata had 800 git clones today (actually 600+ every day).

glensc commented:
from git ... yes they do. you can update documentation from separate PR (i don't have time right now to search all documentation places to change)

ktsaou commented:
closed since PR #1032 adds the required functionality.

glensc commented:
well. it is not the same, this PR added configure option to disable it. yours just "detected" and enabled it unconditionally with no way to override detection. also the compile detection is wrong if you want to support variety of hardware, i.e if you're generating package for distribution (technically not all i585 cpu's support sse, if that matters)

ps: also it's wrong in my opinion to commit generated content to git repo (like those m4 macros), but i already explained this. those autoconf/automake/autoconf-archive packages should be needed only for distribution maintainer and whoever wishes to build straight from git checkout, not released tarballs. i guess you have different priorities than getting things done "right".

but anyway, i'm not going to work on any of this. so the PR can stay closed.

ktsaou commented:
ok, let me understand:

The code I added checks the sse is accepted by the compiler. If I have understood it right, this compiles a tiny c program with these options and if the compilation succeeds it enables these settings.

If the above is right, why will someone need to disable these options?
Of course we could add a configure option to disable them, but please help me understand why this is needed.

Regarding the m4 macros, I am confused. I don't know what you mean by "generated":

1. The `atomic` one is written by me (I couldn't find one).
2. The `generic` one is copied from a patch submitted to autoconf-archives, which I think didn't make it into it (at least I couldn't find it).
3. The `ax_check_compile_flag` (thanks for suggesting it - I took the idea from this PR) is copied directly from autoconf-archives, to avoid having a dependency on autoconf-archives for git users.

So, except the last one that could be removed if we assume autoconf-archives is available everywhere (it is not), the other ones are here to stay.

Am I missing something?

glensc commented:
On 28.09.2016 20:51, Costa Tsaousis wrote:
>
> ok, let me understand:
>
> The code I added checks the sse is accepted by the compiler. If I have 
> understood it right, this compiles a tiny c program with these options 
> and if the compilation succeeds it enables these settings.
>
> If the above is right, why will someone need to disable these options?
> Of course we could add a configure option to disable them, but please 
> help me understand why this is needed.
>

this is compile time decision for that specific cpu. if you run the 
binary produced by those decisions on different cpu, that code will fail 
(program SIGILL) as your decision was compile time, not run time.

so you compile on machine that has sse2 cpu, now the resulting binary 
won't be run on machine without sse2.

amount of such machines (lacking sse2) is probably irrelevant, but i 
bring it out anyway so you could understand. and for (linux) 
distributions intending to support such cpu's lacking sse2, i added 
initially the --disable-sse2 configure flag. so (linux) distribution can 
compile the program on machine having sse2 and allow the resulting 
binary be ran on cpu lacking sse2 support.



> Regarding the m4 macros, I am confused. I don't know what you mean by 
> "generated":
>
>  1. The |atomic| one is written by me (I couldn't find one).
>  2. The |generic| one is copied from a patch submitted to
>     autoconf-archives, which I think didn't make it into it (at least
>     I couldn't find it).
>  3. The |ax_check_compile_flag| (thanks for suggesting it - I took the
>     idea from this PR) is copied directly from autoconf-archives, to
>     avoid having a dependency on autoconf-archives for git users.
>
> So, except the last one that could be removed if we assume 
> autoconf-archives is available everywhere (it is not), the other ones 
> are here to stay.
>

i mean that the m4 macros you copied from autoconf-archive project is 
unnecessary. because it's duplication and can get out of sync 
(autoconf-archive gets fixes, but you forget to update copy in your 
repository) and it's generally not good to stuff such in your code repo. 
your code repo should be about your code :) well. that's my opinion anyway.

-- 
glen

ktsaou commented:
> this is compile time decision for that specific cpu. if you run the
> binary produced by those decisions on different cpu, that code will fail
> (program SIGILL) as your decision was compile time, not run time.

ok, I added option to disable SSE if needed: https://github.com/ktsaou/netdata/commit/9f7d5e3afdd29f66215d6817b63835054516ff05

Regarding the m4 macros, I agree. But, until netdata is distributed by ubuntu, debian, fedora, centos, etc, I think it would be better to keep them in the repo. The distributions can always delete from m4/ the ones found in autoconf-archives.

glensc commented:
also, one point i forgot to bring out about adding files to your code repository from other projects: licensing.

... but luckily both your project and autoconf-archive are `GPL v3+` :)
